src/Microsoft.DotNet.XHarness.CLI/Commands/WASI/Engine/WasiTestCommand.cs

```diff
@@ -0,0 +1,132 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DotNet.XHarness.CLI.CommandArguments.Wasi;
+using Microsoft.DotNet.XHarness.CLI.Commands.Wasm;
+using Microsoft.DotNet.XHarness.Common;
+using Microsoft.DotNet.XHarness.Common.CLI;
+using Microsoft.DotNet.XHarness.Common.Execution;
+using Microsoft.DotNet.XHarness.Common.Logging;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.XHarness.CLI.Commands.Wasi;
+internal class WasiTestCommand : XHarnessCommand<WasiTestCommandArguments>
+{
+    private const string CommandHelp = "Executes tests on WASI using a selected engine";
+
+    protected override WasiTestCommandArguments Arguments { get; } = new();
+    protected override string CommandUsage { get; } = "wasi test [OPTIONS] -- [ENGINE OPTIONS]";
+    protected override string CommandDescription { get; } = CommandHelp;
+
+    public WasiTestCommand() : base(TargetPlatform.WASI, "test", true, new ServiceCollection(), CommandHelp)
+    {
+    }
+
+    protected override async Task<ExitCode> InvokeInternal(ILogger logger)
+    {
+        var processManager = ProcessManagerFactory.CreateProcessManager();
+
+        string engineBinary = Arguments.Engine.Value switch
+        {
+            WasmEngine.WasmTime => "wasmtime",
+            _ => throw new ArgumentException("Engine not set")
+        };
+
+        if (!string.IsNullOrEmpty(Arguments.EnginePath.Value))
+        {
+            engineBinary = Arguments.EnginePath.Value;
+            if (Path.IsPathRooted(engineBinary) && !File.Exists(engineBinary))
+                throw new ArgumentException($"Could not find wasm engine at the specified path - {engineBinary}");
+        }
+      
+        logger.LogInformation($"Using wasm engine {Arguments.Engine.Value} from path {engineBinary}");
+        await PrintVersionAsync(Arguments.Engine.Value.Value, engineBinary);
+
+        var cts = new CancellationTokenSource();
+        try
+        {
+            var engineArgs = new List<string>();
+            engineArgs.Add(Arguments.SubCommand);
+            engineArgs.Add("--dir");
+            engineArgs.Add(".");
+            engineArgs.Add(Arguments.WasmFile);
+            engineArgs.AddRange(Arguments.EngineArgs.Value);
+            engineArgs.AddRange(PassThroughArguments);
+
+            var xmlResultsFilePath = Path.Combine(Arguments.OutputDirectory, "testResults.xml");
+            File.Delete(xmlResultsFilePath);
+
+            var stdoutFilePath = Path.Combine(Arguments.OutputDirectory, "wasi-console.log");
+            File.Delete(stdoutFilePath);
```

Comment:
This also needs the `WasmTestMessagesProcessor` to be used. It will copy all the output to this file, and only show the failures on the console. The symbolicator arguments can be null.